### PR TITLE
Restore wallet using hex

### LINKED ISSUE #767: Allow wallet restore using HEX
Currently GoDCR does not have the option to restore a wallet using the hex private key. Decrediton and dcrwallet both allow this.

### PR DESCRIPTION
**Pull request by crux25**
This PR allows the import of wallet using hex private key. It also shows the hex private key on the wallet backup page.
Closes #767.
restore-page-1
restore-page-2
seed-back-1

### CODE REVIEW

File: ui/page/seedbackup/save_seed.go [hunk 1]
```diff
@@ -173,13 +191,84 @@ func (pg *SaveSeedPage) Layout(gtx C) D {
 						}),
 					)
 				}),
+				layout.Flexed(1, func(gtx C) D {
+					return pg.hexLayout(gtx)
+				}),
```
Comment (Sirmorrison) [inline ui/page/seedbackup/save_seed.go]:
```suggestion
				layout.Flexed(1, pg.hexLayout),
```

File: ui/page/seedbackup/save_seed.go [hunk 2]
```diff
@@ -173,13 +191,84 @@ func (pg *SaveSeedPage) Layout(gtx C) D {
...
+				layout.Rigid(func(gtx C) D {
+					return layout.Inset{Left: values.MarginPadding1}.Layout(gtx, func(gtx C) D { return layout.Dimensions{} })
+				}),
+				layout.Rigid(func(gtx C) D {
+					card.Radius = decredmaterial.CornerRadius{TopRight: 8, TopLeft: 0, BottomRight: 8, BottomLeft: 0}
+					return card.Layout(gtx, pg.copy.Layout)
+				}),
+			)
+		}),
```
Comment (Sirmorrison) [inline ui/page/seedbackup/save_seed.go]:
```suggestion
		layout.Rigid(func(gtx C) D {
			return layout.Flex{Alignment: layout.Middle}.Layout(gtx,
				layout.Rigid(func(gtx C) D {
					card.Radius = decredmaterial.CornerRadius{TopRight: 0, TopLeft: 8, BottomRight: 0, BottomLeft: 8}

					return card.Layout(gtx, func(gtx C) D {
						return layout.UniformInset(values.MarginPadding16).Layout(gtx, func(gtx layout.Context) layout.Dimensions {
							seedString := pg.seed
							if seedString != "" {
								hexString, _ := load.SeedWordsToHex(pg.seed)
								pg.hexLabel.Text = hexString
							}
							return pg.hexLabel.Layout(gtx)
						})
					})
				}),
				layout.Rigid(func(gtx C) D {
					card.Radius = decredmaterial.CornerRadius{TopRight: 8, TopLeft: 0, BottomRight: 8, BottomLeft: 0}
					return layout.Inset{Left: values.MarginPadding1}.Layout(gtx, func(gtx C) D { 
						return card.Layout(gtx, pg.copy.Layout)
					})
				}),
			)
		}),
```

**Review by Sirmorrison — CHANGES_REQUESTED** [review]
Kindly provide a test HEX code.

image

So i have my reservation to this design as we have to be production ready with every commit to master.

That said, when a user decides to restore with hex, this should open a modal (text input moda), that takes the hex rather than the center aligned editor widget used.

image

I am so sure why we need to validate the hex like this, however, i believe the hex validation can be done in the background and error from validation displayed on the UI of the modal.

**Comment by crux25** [comment]
> Kindly provide a test HEX code.
> 
> image
> 
> So i have my reservation to this design as we have to be production ready with every commit to master.
> 
> That said, when a user decides to restore with hex, this should open a modal (text input moda), that takes the hex rather than the center aligned editor widget used.
> 
> image
> 
> I am so sure why we need to validate the hex like this, however, i believe the hex validation can be done in the background and error from validation displayed on the UI of the modal.

Create a wallet. Try to back up it's seed. You'll find the wallet's hex code on the seed backup page.

**Comment by crux25** [comment]
You said you have your reservation to the design? Can you list the reservations so I can fix it and have it production ready by next review. So if I understand you correctly, I should maintain the tabs but instead of showing the current hex page I should use a modal instead?

**Comment by Sirmorrison** [comment]
> You said you have your reservation to the design? Can you list the reservations so I can fix it and have it production ready by next review. So if I understand you correctly, I should maintain the tabs but instead of showing the current hex page I should use a modal instead?

This can work or we could make use of buttons that performs different actions like what we have on the privacy page for both manual setup and automatic setup.

Manual setup redirects to a page while the automatic opens up a modal.

**Comment by dreacot** [comment]
I tried to restore an already existing wallet using hex, i first got the "wallet with identical seed error" error after inputting the hex code into the modal, but i was still presented with a modal for inputting wallet name and password.

I was ultimately able to create a wallet using the hex for a wallet that already existed

File: ui/modal/hex_wallet_restore_modal.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	return hm.randomID
+}
+
+func (hm *HexRestoreModal) OnResume() {
+	hm.hexEditor.Editor.Focus()
+}
+
+func (hm *HexRestoreModal) OnDismiss() {
+
+}
```
Comment (dreacot) [inline ui/modal/hex_wallet_restore_modal.go]:
```suggestion
func (hm *HexRestoreModal) OnDismiss() {}
```

**Comment by crux25** [comment]
> I tried to restore an already existing wallet using hex, i first got the "wallet with identical seed error" error after inputting the hex code into the modal, but i was still presented with a modal for inputting wallet name and password.
> 
> I was ultimately able to create a wallet using the hex for a wallet that already existed

I'll verify this. I'll run a test and comment in a bit.

File: ui/modal/hex_wallet_restore_modal.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	}
+
+	if hm.customWidget != nil {
+		w = append(w, hm.customWidget)
+	}
+
+	w = append(w, editor)
+	w = append(w, actionButtons)
+
+	return hm.modal.Layout(gtx, w)
```
Comment (Sirmorrison) [inline ui/modal/hex_wallet_restore_modal.go]: We dont need this modal. We already have the `text input modal` to handle scenarios of this nature.

File: ui/load/utils.go [hunk 5]
```diff
@@ -15,3 +24,56 @@ const (
...
+
+	seedHex = hex.EncodeToString(seedByte)
+	return seedHex, nil
+}
+
+// checksumByte returns the checksum byte used at the end of the seed mnemonic
+// encoding.  The "checksum" is the first byte of the double SHA256.
+func checksumByte(data []byte) byte {
+	intermediateHash := sha256.Sum256(data)
+	return sha256.Sum256(intermediateHash[:])[0]
```
Comment (Sirmorrison) [inline ui/load/utils.go]: Let move this lets move these lines of code to `components package utils.go`

**Review by Sirmorrison — CHANGES_REQUESTED** [review]
Will test and review further after these are resolved.

File: ui/page/seedbackup/save_seed.go [hunk 6]
```diff
@@ -42,13 +47,27 @@ func NewSaveSeedPage(l *load.Load, wallet *dcrlibwallet.Wallet) *SaveSeedPage {
...
+	pg.hexLabel.MaxLines = 1
+	pg.copy.Background = color.NRGBA{}
+	pg.copy.HighlightColor = pg.Theme.Color.SurfaceHighlight
+	pg.copy.Color = pg.Theme.Color.Primary
+	pg.copy.Inset = layout.Inset{
+		Top:    values.MarginPadding16,
+		Bottom: values.MarginPadding16,
+		Left:   values.MarginPadding16,
+		Right:  values.MarginPadding16,
+	}
```
Comment (Sirmorrison) [inline ui/page/seedbackup/save_seed.go]:
```suggestion
	pg.copy.Inset = layout.UniformInset(values.MarginPadding16)
```

File: ui/page/seedbackup/save_seed.go [hunk 7]
```diff
@@ -173,13 +191,84 @@ func (pg *SaveSeedPage) Layout(gtx C) D {
...
+				}),
+				layout.Rigid(func(gtx C) D {
+					return layout.Inset{Left: values.MarginPadding1}.Layout(gtx, func(gtx C) D { return layout.Dimensions{} })
+				}),
+				layout.Rigid(func(gtx C) D {
+					card.Radius = decredmaterial.CornerRadius{TopRight: 8, TopLeft: 0, BottomRight: 8, BottomLeft: 0}
+					return card.Layout(gtx, pg.copy.Layout)
+				}),
+			)
+		}),
```
Comment (Sirmorrison) [inline ui/page/seedbackup/save_seed.go]:
```suggestion
		layout.Rigid(func(gtx C) D {
			return layout.Flex{Alignment: layout.Middle}.Layout(gtx,
				layout.Rigid(func(gtx C) D {
					card.Radius = decredmaterial.CornerRadius{TopRight: 0, TopLeft: 8, BottomRight: 0, BottomLeft: 8}
					return card.Layout(gtx, func(gtx C) D {
						return layout.UniformInset(values.MarginPadding16).Layout(gtx, func(gtx layout.Context) layout.Dimensions {
							seedString := pg.seed
							if seedString != "" {
								hexString, _ := load.SeedWordsToHex(pg.seed)
								pg.hexLabel.Text = hexString
							}
							return pg.hexLabel.Layout(gtx)
						})
					})
				}),
				layout.Rigid(func(gtx C) D {
					card.Radius = decredmaterial.CornerRadius{TopRight: 8, TopLeft: 0, BottomRight: 8, BottomLeft: 0}
					return layout.Inset{Left: values.MarginPadding1}.Layout(gtx, func(gtx C) D { 
						return card.Layout(gtx, pg.copy.Layout)
					})
				}),
			)
		}),
```

File: ui/page/wallets/restore_page.go [hunk 8]
```diff
@@ -130,317 +67,181 @@ func (pg *Restore) Layout(gtx layout.Context) layout.Dimensions {
...
-		return pg.validateSeed.Layout(gtx)
+func (pg *Restore) restoreLayout(gtx layout.Context) layout.Dimensions {
+	return components.UniformPadding(gtx, func(gtx C) D {
+		return layout.Flex{Axis: layout.Vertical}.Layout(gtx,
+			layout.Rigid(pg.tabLayout),
+			layout.Rigid(pg.Theme.Separator().Layout),
+			layout.Flexed(1, func(gtx C) D {
+				return layout.Inset{Top: values.MarginPadding5}.Layout(gtx, func(gtx C) D {
+					return pg.indexLayout(gtx)
+				})
```
Comment (Sirmorrison) [inline ui/page/wallets/restore_page.go]:
```suggestion
				return layout.Inset{Top: values.MarginPadding5}.Layout(gtx, pg.indexLayout)
```

**Review by Sirmorrison — CHANGES_REQUESTED** [review]
functionality looks good. just for the requested changes

**Comment by dreacot** [comment]
1. I chose to restore using HEX;
Screenshot from 2022-06-02 16-35-01

when i got to the above screen, i decided to cancel the process.

Now i'm unable to select hex again until i first switch to seed words tab
Screenshot from 2022-06-02 16-35-56

2. When I put in an invalid HEX, the modal closes and shows a Toast. The expected behavior is for the modal to remain open and the input field should notify of the error
Screenshot from 2022-06-02 16-37-07.

See example below
Screenshot from 2022-06-02 16-38-25

File: ui/page/components/utils.go [hunk 9]
```diff
@@ -37,3 +42,56 @@ func RetryFunc(retryAttempts int, sleepDur time.Duration, funcDesc string, errFu
...
+	if len(seedByte) < minSeedBytes || len(seedByte) > maxSeedBytes {
+		return seedHex, fmt.Errorf("Invalid seed bytes length")
+	}
+
+	seedHex = hex.EncodeToString(seedByte)
+	return seedHex, nil
+}
+
+// checksumByte returns the checksum byte used at the end of the seed mnemonic
+// encoding.  The "checksum" is the first byte of the double SHA256.
```
Comment (dreacot) [inline ui/page/components/utils.go]:
```suggestion
// encoding. The "checksum" is the first byte of the double SHA256.
```

File: ui/page/wallets/seed_restore_page.go [hunk 10]
```diff
@@ -0,0 +1,610 @@
+package wallets
+
+import (
+	"fmt"
+
```
Comment (Sirmorrison) [inline ui/page/wallets/seed_restore_page.go]: remove white space

File: ui/page/seedbackup/save_seed.go [hunk 11]
```diff
@@ -173,13 +186,76 @@ func (pg *SaveSeedPage) Layout(gtx C) D {
...
+func (pg *SaveSeedPage) handleCopyEvent(gtx layout.Context) {
+	if pg.copy.Clicked() {
+		clipboard.WriteOp{Text: pg.hexLabel.Text}.Add(gtx.Ops)
+
+		pg.copy.Text = "Copied!"
+		pg.copy.Color = pg.Theme.Color.Success
+		time.AfterFunc(time.Second*3, func() {
+			pg.copy.Text = "Copy"
+			pg.copy.Color = pg.Theme.Color.Primary
+		})
```
Comment (Sirmorrison) [inline ui/page/seedbackup/save_seed.go]:
```suggestion
		time.AfterFunc(time.Second*3, func() {
			pg.copy.Text = "Copy"
			pg.copy.Color = pg.Theme.Color.Primary
			pg.RefreshWindow()
		})
```

After 3 secs you need to refresh the window for the changes to be displayed.

**Review by Sirmorrison — CHANGES_REQUESTED** [review]

**Review by Sirmorrison — CHANGES_REQUESTED** [review]
https://user-images.githubusercontent.com/27733432/173023870-14fcd5db-199e-4837-a08c-5b932105212a.mov

* so i am noticing some unexpected behaviors.. How are hex validated? are we just checking for the length because if i change the last `n digits` of the hex, it still imports a wallet.
* Why does errors from validating the hex minimize the modal eg (invalid hex)? the errors should be displayed under hex input editor.
* lastly, when i i have input a wrong hex (ie editing the original hex key values), it opens up the wallet editor.. if i click on cancel to minimize the wallet creation modal, clicking on the hex nav button again, does not open the modal.

**Comment by crux25** [comment]
> 



> Screen.Recording.2022-06-10.at.09.21.53.mov
> 
>     * so i am noticing some unexpected behaviors.. How are hex validated? are we just checking for the length because if i change the last `n digits` of the hex, it still imports a wallet.
> 
>     * Why does errors from validating the hex minimize the modal eg (invalid hex)? the errors should be displayed under hex input editor.
> 
>     * lastly, when i i have input a wrong hex (ie editing the original hex key values), it opens up the wallet editor.. if i click on cancel to minimize the wallet creation modal, clicking on the hex nav button again, does not open the modal.

Hex validation is handled by dcrwallet basically it checks that the supplied hex is valid; that is the character contains [0-9][a-f], it also checks the hex length. If you're interested in the implementation you can look at https://github.com/decred/dcrwallet/blob/4ed5a4c677f9bdb1b8e70ceccf476fbff413cd79/walletseed/seed.go#L71

**Comment by crux25** [comment]
> 



> Screen.Recording.2022-06-10.at.09.21.53.mov
> 
>     * so i am noticing some unexpected behaviors.. How are hex validated? are we just checking for the length because if i change the last `n digits` of the hex, it still imports a wallet.
> 
>     * Why does errors from validating the hex minimize the modal eg (invalid hex)? the errors should be displayed under hex input editor.
> 
>     * lastly, when i i have input a wrong hex (ie editing the original hex key values), it opens up the wallet editor.. if i click on cancel to minimize the wallet creation modal, clicking on the hex nav button again, does not open the modal.

Hex validation is handled by dcrwallet basically it checks that the supplied hex is valid; that is the character contains [0-9][a-f], it also checks the hex length. If you're interested in the implementation you can look at https://github.com/decred/dcrwallet/blob/4ed5a4c677f9bdb1b8e70ceccf476fbff413cd79/walletseed/seed.go#L71

**Review by Sirmorrison — APPROVED** [review]

**Pull request merged**